Libplanet/Blockchain/Policies/VolatileStagePolicy.cs

```diff
@@ -0,0 +1,152 @@
+#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using Libplanet.Action;
+using Libplanet.Tx;
+
+namespace Libplanet.Blockchain.Policies
+{
+    /// <summary>
+    /// In-memory staged transactions.
+    /// </summary>
+    /// <typeparam name="T">An <see cref="IAction"/> type.  It should match
+    /// to <see cref="BlockChain{T}"/>'s type parameter.</typeparam>
+    public class VolatileStagePolicy<T> : IStagePolicy<T>
+        where T : IAction, new()
+    {
+        private readonly ConcurrentDictionary<TxId, Transaction<T>> _set;
+
+        private readonly List<TxId> _queue;
+
+        private readonly ReaderWriterLockSlim _lock;
+
+        /// <summary>
+        /// Creates a new <see cref="VolatileStagePolicy{T}"/> instance.
+        /// </summary>
+        public VolatileStagePolicy()
+        {
+            _set = new ConcurrentDictionary<TxId, Transaction<T>>();
+            _queue = new List<TxId>();
+            _lock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+        }
+
+        /// <inheritdoc cref="IStagePolicy{T}.Stage(BlockChain{T}, Transaction{T})"/>
+        public void Stage(BlockChain<T> blockChain, Transaction<T> transaction)
+        {
+            bool stageAgain = !_set.TryAdd(transaction.Id, transaction);
+            if (stageAgain)
+            {
+                bool staged;
+                _lock.EnterUpgradeableReadLock();
+                try
+                {
+                    staged = _queue.Contains(transaction.Id);
+                }
+                catch (Exception)
+                {
+                    _lock.ExitUpgradeableReadLock();
+                    throw;
+                }
+
+                if (staged)
+                {
+                    _lock.ExitUpgradeableReadLock();
+                    return;
+                }
+            }
+
+            _lock.EnterWriteLock();
+            try
+            {
+                _queue.Add(transaction.Id);
+            }
+            finally
+            {
+                _lock.ExitWriteLock();
+                if (stageAgain)
+                {
+                    _lock.ExitUpgradeableReadLock();
+                }
+            }
```

Comment:
How about this?

```suggestion
            try
            {
                if (!_set.TryAdd(transaction.Id, transaction))
                {
                    _lock.EnterUpgradeableReadLock();

                    if (_queue.Contains(transaction.Id))
                    {
                        return;
                    }
                }

                _lock.EnterWriteLock();
                _queue.Add(transaction.Id);
            }
            finally
            {
                _lock.ExitWriteLock();
                if (_lock.IsUpgradableReadLockHeld)
                {
                    _lock.ExitUpgradeableReadLock();
                }
            }
```

Or, I guess it can be fine if we always hold `EnterUpgradeableReadLock` like below

```csharp
            try
            {
                _lock.EnterUpgradeableReadLock();
                if (!_set.TryAdd(transaction.Id, transaction) && _queue.Contains(transaction.Id))
                {
                    return;
                }

                _lock.EnterWriteLock();
                try
                {
                    _queue.Add(transaction.Id);
                }
                finally
                {
                    _lock.ExitWriteLock();
                }
            }
            finally
            {
                _lock.ExitUpgradeableReadLock();
            }

```

Reply:
What I intended is that `_lock` is only for `_queue` as `_set` in itself is a `ConcurrentDictionary<TxId, Transaction<T>>`.  Do you think both `_queue` and `_set` should be locked?

Reply:
> the first suggestion might be a problem when any exception is thrown before _lock.EnterWriteLock() is reached.

Right. then...

```suggestion
            try
            {
                if (!_set.TryAdd(transaction.Id, transaction))
                {
                    _lock.EnterUpgradeableReadLock();

                    if (_queue.Contains(transaction.Id))
                    {
                        return;
                    }
                }

                _lock.EnterWriteLock();
                try
                {
                    _queue.Add(transaction.Id);
                }
                finally
                {
                   _lock.ExitWriteLock();
                }
            }
            finally
            {
                if (_lock.IsUpgradableReadLockHeld)
                {
                    _lock.ExitUpgradeableReadLock();
                }
            }
```

> The second suggestion seems more coarse-grained than before.

It's also right, but IMHO, it doesn't affect the parallelism seriously. 😕

Reply:
@longfin Applied your suggestion: 97263d76c.